timings/state.go

```diff
@@ -121,3 +129,37 @@ func (t *Timings) Save(st *state.State) {
 	}
 	st.Set("timings", stateTimings)
 }
+
+// Get returns timings for which filter predicate is true and filters out nested timings whose level is greater than maxLevel.
+// Negative maxLevel value disables filtering by level.
+// It's responsibility of the caller to lock the state before calling this function.
+func Get(st *state.State, maxLevel int, filter func(tags map[string]string) bool) ([]*TimingsInfo, error) {
+	var stateTimings []rootTimingsJSON
+	if err := st.Get("timings", &stateTimings); err != nil && err != state.ErrNoState {
+		return nil, fmt.Errorf("could not get timings data from the state: %v", err)
+	}
+
+	var result []*TimingsInfo
+	for _, tm := range stateTimings {
+		if !filter(tm.Tags) {
+			continue
+		}
+		res := &TimingsInfo{
+			Tags: tm.Tags,
+		}
+		if maxLevel < 0 { // no level-based filtering?
+			res.NestedTimings = tm.NestedTimings
```

Comment:
There's an inconsistency here in that ,  res is appended even if NestedTimings has len() == 0 which is different from the code below?

Reply:
Thanks, I've slightly changed this code. In fact Save() should not store timings if it has no nested spans, I made the fix to Save. Then here we always have at least one nested timing.